### PR TITLE
check if the markdown object is visible 

### PR DESCRIPTION
**Pull request by OsmondJiang**
this will help validation library to determine if the validated node is visible  

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5874)

### CODE REVIEW

File: test/docfx.Test/lib/MarkdigUtilityTest.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        [InlineData("<a>test</a>", true)]
+        [InlineData("`code`", true)]
+
+        [InlineData("<!--comments-->", false)]
+        [InlineData("    ", false)]
+        [InlineData("  \n  \n  ", false)]
+        [InlineData("  \n\n  \n\n  ", false)]
+        [InlineData("  \n <!--comments--> \n  ", false)]
+        [InlineData("[](https://github.com)", false)]
+        [InlineData("<a></a>", false)]
```
Comment (fkpwolf) [inline test/docfx.Test/lib/MarkdigUtilityTest.cs:30]: so we didn't check css like `<div style="display:none;">invisible indeed</div>`

File: test/docfx.Test/lib/MarkdigUtilityTest.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        [InlineData("[link](https://github.com)", true)]
+        [InlineData("![image](image.png)", true)]
+        [InlineData("<img src=\"imag.png\"/>", true)]
+        [InlineData("<img src=\"imag.png\">abc</img>", true)]
+        [InlineData("<a href=\"www.microsoft.com\"><img src=\"imag.png\"/></a>", true)]
+        [InlineData("<a>test</a>", true)]
+        [InlineData("`code`", true)]
+
+        [InlineData("<!--comments-->", false)]
+        [InlineData("    ", false)]
```
Comment (XiaofuHuang) [inline test/docfx.Test/lib/MarkdigUtilityTest.cs:25]: Would you like to add a test case about "&nbsp;" and TripleColonBlock of image?

File: test/docfx.Test/lib/MarkdigUtilityTest.cs (hunk 1 above)
Comment (OsmondJiang) [reply test/docfx.Test/lib/MarkdigUtilityTest.cs:30]: style attribute is disallowed in docfx

File: test/docfx.Test/lib/MarkdigUtilityTest.cs (hunk 2 above)
Comment (OsmondJiang) [reply test/docfx.Test/lib/MarkdigUtilityTest.cs:25]: ok, i will add

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 3]
```diff
@@ -153,6 +154,56 @@ public static string ToPlainText(this MarkdownObject containerBlock)
             return writer.ToString();
         }
 
+        public static string ToHtml(this MarkdownObject markdownObject)
```
Comment (yufeih) [inline src/docfx/lib/markdown/MarkdigUtility.cs:157]: This way of check visibility is very likely to have visible perf impact if applied to every node (Equivalent to render markdown AST twice).

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 4]
```diff
@@ -153,6 +154,69 @@ public static string ToPlainText(this MarkdownObject containerBlock)
             return writer.ToString();
         }
 
+        public static bool IsVisible(this MarkdownObject markdownObject)
+        {
+            var visiable = false;
+            Visit(markdownObject, node =>
+            {
+                if (node is HtmlBlock htmlBlock)
```
Comment (yufeih) [inline src/docfx/lib/markdown/MarkdigUtility.cs:162]: nit: consider `switch case`?
```suggestion
                if (node is HtmlBlock htmlBlock)
```

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 5]
```diff
@@ -153,6 +154,69 @@ public static string ToPlainText(this MarkdownObject containerBlock)
...
+                {
+                    visiable = visiable || VisibleHtml(htmlBlock.Lines.ToString());
```
Comment (yufeih) [inline src/docfx/lib/markdown/MarkdigUtility.cs:164]: Aggregate the visibility of each line to save the `ToString` allocation?

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 6]
```diff
@@ -153,6 +154,69 @@ public static string ToPlainText(this MarkdownObject containerBlock)
...
+
+            bool VisibleHtml(string html)
+            {
+                var reader = new HtmlReader(html);
+                while (reader.Read(out var token))
+                {
+                    // check if there are text
+                    if (token.Type == HtmlTokenType.Text)
+                    {
+                        if (!string.IsNullOrWhiteSpace(token.RawText.ToString()))
```
Comment (yufeih) [inline src/docfx/lib/markdown/MarkdigUtility.cs:200]: Save the `ToString()` allocation

```suggestion
                        if (!token.RawText.IsWhiteSpace())
```

Comment (OsmondJiang) [reply src/docfx/lib/markdown/MarkdigUtility.cs:200]: can't find the IsWhiteSpace extension method

**Comment by mingwli** [comment]
The [build](https://dev.azure.com/ceapex/Engineering/_build/results?buildId=146150&view=results) has actually already passed.
image

**Pull request merged**